packages/service/src/utils/document-history.ts

```diff
@@ -2,37 +2,84 @@ import * as Y from "yjs";
 import { pool, db } from "../db";
 import gitDiff from "git-diff";
 
-function fetchDocumentHistory(document_id: string) {
-  return db
-    .select(
-      "document_updates_queue",
-      { document_id },
-      {
-        order: { by: "created_at", direction: "ASC" },
-        lateral: {
-          user: db.selectExactlyOne(
-            "users",
-            { id: db.parent("user_id") },
-            { columns: ["id", "name"] }
-          ),
-        },
-      }
-    )
-    .run(pool);
+//select changes in document_updates_queue
+export async function fetchDocumentHistoryBuckets(document_id: string){
+  const buckets = await db.sql`
+  SELECT to_timestamp (extract ('epoch' from created_at)::int/60*60) as timeslice, user_id, document_id, users.name as username, array_agg(document_update)
+       from document_updates_queue left join users on document_updates_queue.user_id = users.id
+       Where document_id = ${db.param(document_id)}
+       group by document_id, timeslice, user_id, users.name
+       order by timeslice ASC`.run(pool);
+
+  return buckets;
+}
+
+//drop histories with document id
+export async function removeFromDocumentHistory(document){
+  await db.sql`
+    DELETE FROM document_history
+    where document_history.document_id = ${db.param(document.document_id)}
+  `.run(pool);
+}
+//put into one transaction
+const dropAndAdd = (document, arr) =>
+  db.serializable(pool, txnClient => Promise.all([
+    db.deletes('document_history',
+    {
+      document_id: document.id
+    }).run(txnClient),
+    db.insert('document_history',
+      arr
+    ).run(txnClient),
+  ]));
+
+
+
+export async function insertIntoDocumentHistory(document){
+  
+  let arr = await buildDocumentHistoryBuckets(document);
+
+  let insertArr = [];
+  for (let i = 0; i < arr.length; i++){
+    insertArr[i] = {document_id: document.id, user_id: (arr[i].user.id.length != 0) ? arr[i].user.id: null, diff: arr[i].diff};
```

Comment:
Can be simplified as

```ts
const insertArr = arr.map(history => ({
  document_id: document.id, 
  user_id: (history.user.id.length != 0) 
    ? history.user.id
    : null, 
  diff: history.diff
}));

Reply:
but it would be better to just push this into `buildDocumentHistoryBuckets`, which is already looping through all of the history items anyway, no need to do another loop here.